apps/ejabberd/src/gen_mod.erl

```diff
@@ -50,9 +50,9 @@
 
 -include("ejabberd.hrl").
 
--record(ejabberd_module, {module_host :: ejabberd:server(),
-                          opts :: list()
-                         }).
+-record(ejabberd_module, {module_host, opts}).
+-type ejabberd_module() :: #ejabberd_module{module_host :: {module(), ejabberd:server()},
```

Comment:
Because of match spec construction a dozen of lines down, i.e. this:

```erlang
gen_mod.erl:325: Record construction #ejabberd_module{module_host::{'$1',_},opts::'_'} violates the declared type of field opts::'undefined' | [any()]
```

I admit it's not ideal, because now one might actually construct an invalid record and Dialyzer won't complain. On the other hand, if an input parameter towards a function is specced properly (that is with `ejabberd_module()`, not `#ejabberd_module{}`) then Dialyzer will complain about passing around such invalid records.